### ISSUE TITLE
Changes to metrics passed to on_batch_end(...)

### ISSUE THREAD

**Issue by leland-hepworth** [tensorflow/tensorflow#58098]
<details><summary>Click to expand!</summary> 
 
 ### Issue Type

Bug

### Source

source

### Tensorflow Version

tf 2.2+

### Custom Code

No

### OS Platform and Distribution

Linux Ubuntu 22.04.1

### Mobile device

Linux Ubuntu 22.04.1

### Python version

3.7.13

### CUDA/cuDNN version

11.7

### Current Behaviour?

I recently updated code from TensorFlow 1 and was getting strange results for my callbacks. I eventually found [issue #36400](https://github.com/tensorflow/tensorflow/issues/36400) and decided to test that. I discovered that in tf 2.2+ on_batch_end is now provided a moving average of all losses so far in the current epoch, while in previous versions only the loss for the current batch is provided.

This matters when using the learning rate finder callback described in [Cyclical Learning Rates for Training Neural Networks by Leslie N. Smith](https://arxiv.org/pdf/1506.01186.pdf).  When using the moving average losses, the suggested range of learning rate values is shifted to the right, resulting in learning rates that are much higher than the optimal range, and leading to poorer performance than when the loss for only the current batch is used correctly.

### Standalone code to reproduce the issue

```python
import sys

from tensorflow import keras
from tensorflow.keras import layers

EPOCHS = 3
STEPS_PER_EPOCH = 10
BATCH_SIZE = 64
SAMPLES = BATCH_SIZE * STEPS_PER_EPOCH


inputs = keras.Input(shape=(784,), name='digits')
x = layers.Dense(4*4096, activation='relu', name='dense_1')(inputs)
x = layers.Dense(4*4096, activation='relu', name='dense_2')(x)
x = layers.Dense(4*4096, activation='relu', name='dense_3')(x)
outputs = layers.Dense(10, activation='softmax', name='predictions')(x)

model = keras.Model(inputs=inputs, outputs=outputs)
(x_train, y_train), _ = keras.datasets.mnist.load_data()
x_train = x_train[:SAMPLES].reshape(SAMPLES, 784).astype('float32') / 255
y_train = y_train[:SAMPLES].astype('float32')

class PrintStatsCallback(keras.callbacks.Callback):
    def on_batch_end(self, batch, logs=None):
        print('\nON_BATCH_END: loss: {loss:.4f} acc: {accuracy:.4f}'.format_map(logs), file=sys.stderr)
    def on_epoch_end(self, epoch, logs=None):
        print('\nON_EPOCH_END: loss: {loss:.4f} acc: {accuracy:.4f}'.format_map(logs), file=sys.stderr)

# Specify the training configuration (optimizer, loss, metrics)
model.compile(optimizer=keras.optimizers.SGD(lr=0.1),
              loss='sparse_categorical_crossentropy', metrics=['accuracy'])

model.fit(x_train, y_train,
          epochs=EPOCHS,
          steps_per_epoch=STEPS_PER_EPOCH,
          batch_size=BATCH_SIZE,
          callbacks=[PrintStatsCallback()])
```

### Relevant log output

```shell
Epoch 1/3

ON_BATCH_END: loss: 2.3025 acc: 0.0938
 1/10 [==>...........................] - ETA: 0s - loss: 2.3025 - accuracy: 0.0938
ON_BATCH_END: loss: 2.2634 acc: 0.1484
 2/10 [=====>........................] - ETA: 17s - loss: 2.2634 - accuracy: 0.1484
ON_BATCH_END: loss: 2.2258 acc: 0.1771
 3/10 [========>.....................] - ETA: 19s - loss: 2.2258 - accuracy: 0.1771
ON_BATCH_END: loss: 2.1891 acc: 0.2656
 4/10 [===========>..................] - ETA: 19s - loss: 2.1891 - accuracy: 0.2656
ON_BATCH_END: loss: 2.1599 acc: 0.3031
 5/10 [==============>...............] - ETA: 16s - loss: 2.1599 - accuracy: 0.3031
ON_BATCH_END: loss: 2.1139 acc: 0.3516
 6/10 [=================>............] - ETA: 14s - loss: 2.1139 - accuracy: 0.3516
ON_BATCH_END: loss: 2.0604 acc: 0.4062
 7/10 [====================>.........] - ETA: 10s - loss: 2.0604 - accuracy: 0.4062
ON_BATCH_END: loss: 2.0273 acc: 0.4258
 8/10 [=======================>......] - ETA: 7s - loss: 2.0273 - accuracy: 0.4258
ON_BATCH_END: loss: 2.0036 acc: 0.4306
 9/10 [==========================>...] - ETA: 3s - loss: 2.0036 - accuracy: 0.4306
ON_BATCH_END: loss: 1.9565 acc: 0.4547
10/10 [==============================] - ETA: 0s - loss: 1.9565 - accuracy: 0.4547
ON_EPOCH_END: loss: 1.9565 acc: 0.4547
10/10 [==============================] - 38s 4s/step - loss: 1.9565 - accuracy: 0.4547
Epoch 2/3

ON_BATCH_END: loss: 1.4679 acc: 0.7969
 1/10 [==>...........................] - ETA: 0s - loss: 1.4679 - accuracy: 0.7969
ON_BATCH_END: loss: 1.3991 acc: 0.7734
 2/10 [=====>........................] - ETA: 16s - loss: 1.3991 - accuracy: 0.7734
ON_BATCH_END: loss: 1.3518 acc: 0.7604
 3/10 [========>.....................] - ETA: 19s - loss: 1.3518 - accuracy: 0.7604
ON_BATCH_END: loss: 1.2953 acc: 0.7852
 4/10 [===========>..................] - ETA: 19s - loss: 1.2953 - accuracy: 0.7852
ON_BATCH_END: loss: 1.2575 acc: 0.7906
 5/10 [==============>...............] - ETA: 16s - loss: 1.2575 - accuracy: 0.7906
ON_BATCH_END: loss: 1.2541 acc: 0.7708
 6/10 [=================>............] - ETA: 14s - loss: 1.2541 - accuracy: 0.7708
ON_BATCH_END: loss: 1.2084 acc: 0.7701
 7/10 [====================>.........] - ETA: 10s - loss: 1.2084 - accuracy: 0.7701
ON_BATCH_END: loss: 1.1886 acc: 0.7578
 8/10 [=======================>......] - ETA: 7s - loss: 1.1886 - accuracy: 0.7578
ON_BATCH_END: loss: 1.1727 acc: 0.7587
 9/10 [==========================>...] - ETA: 3s - loss: 1.1727 - accuracy: 0.7587
ON_BATCH_END: loss: 1.1418 acc: 0.7609
10/10 [==============================] - ETA: 0s - loss: 1.1418 - accuracy: 0.7609
ON_EPOCH_END: loss: 1.1418 acc: 0.7609
10/10 [==============================] - 38s 4s/step - loss: 1.1418 - accuracy: 0.7609
Epoch 3/3

ON_BATCH_END: loss: 0.7979 acc: 0.8125
 1/10 [==>...........................] - ETA: 0s - loss: 0.7979 - accuracy: 0.8125
ON_BATCH_END: loss: 0.8240 acc: 0.7812
 2/10 [=====>........................] - ETA: 16s - loss: 0.8240 - accuracy: 0.7812
ON_BATCH_END: loss: 0.8083 acc: 0.7865
 3/10 [========>.....................] - ETA: 19s - loss: 0.8083 - accuracy: 0.7865
ON_BATCH_END: loss: 0.8098 acc: 0.7891
 4/10 [===========>..................] - ETA: 19s - loss: 0.8098 - accuracy: 0.7891
ON_BATCH_END: loss: 0.8060 acc: 0.7781
 5/10 [==============>...............] - ETA: 16s - loss: 0.8060 - accuracy: 0.7781
ON_BATCH_END: loss: 0.8370 acc: 0.7578
 6/10 [=================>............] - ETA: 14s - loss: 0.8370 - accuracy: 0.7578
ON_BATCH_END: loss: 0.8413 acc: 0.7411
 7/10 [====================>.........] - ETA: 10s - loss: 0.8413 - accuracy: 0.7411
ON_BATCH_END: loss: 0.8072 acc: 0.7500
 8/10 [=======================>......] - ETA: 7s - loss: 0.8072 - accuracy: 0.7500
ON_BATCH_END: loss: 0.7843 acc: 0.7604
 9/10 [==========================>...] - ETA: 3s - loss: 0.7843 - accuracy: 0.7604
ON_BATCH_END: loss: 0.7861 acc: 0.7594
10/10 [==============================] - ETA: 0s - loss: 0.7861 - accuracy: 0.7594
ON_EPOCH_END: loss: 0.7861 acc: 0.7594
10/10 [==============================] - 38s 4s/step - loss: 0.7861 - accuracy: 0.7594
```
</details>

**Comment by leland-hepworth** [comment]
Also the "Current Behaviour" section in the issue template is being formatted as code.  Here it is formatted as a comment:

I recently updated code from TensorFlow 1 and was getting strange results for my callbacks. I eventually found [issue #36400](https://github.com/tensorflow/tensorflow/issues/36400) and decided to test that. I discovered that in tf 2.2+ on_batch_end is now provided a moving average of all losses so far in the current epoch, while in previous versions only the loss for the current batch is provided.

This matters when using the learning rate finder callback described in [Cyclical Learning Rates for Training Neural Networks by Leslie N. Smith](https://arxiv.org/pdf/1506.01186.pdf).  When using the moving average losses, the suggested range of learning rate values is shifted to the right, resulting in learning rates that are much higher than the optimal range, and leading to poorer performance than when the loss for only the current batch is used correctly.

**Comment by leland-hepworth** [comment]
In the meantime, I've added the following code to my learning rate finder callback to get the previous functionality:

```python
def __init__(self, ... , is_average_loss=True):
    # ...
    self.is_average_loss = is_average_loss

def on_epoch_begin(self, epoch, logs={}):
    if self.is_average_loss:
        self.previous_loss_sum = 0

def on_train_batch_end(self, batch, logs={}):
    if self.is_average_loss:
        current_loss_sum =  (batch + 1) * logs['loss']
        loss = current_loss_sum - self.previous_loss_sum
        self.previous_loss_sum = current_loss_sum
    else:
        loss = logs['loss']
    
    # ...
```

**Comment by tiruk007** [comment]
Hi @leland-hepworth 
I was able to reproduce the issue on colab using TF v2.10. Please find the gist [here](https://colab.research.google.com/gist/tiruk007/b119e925e202d52a2e7ea5d1b1690f86/58098.ipynb) for reference.

This issue seems to be Keras issue. Please post this issue on [keras-team/keras repo.](https://github.com/keras-team/keras/issues) All issues and PRs related to keras will be addressed in that repo.
To know more see;
https://discuss.tensorflow.org/t/keras-project-moved-to-new-repository-in-https-github-com-keras-team-keras/1999.

Thank you!

**Comment by leland-hepworth** [comment]
@tiruk007, thanks for your reply!

I created [keras issue #17167](https://github.com/keras-team/keras/issues/17167), so this can be closed.
